src/Lecture2.hs

```diff
@@ -158,16 +178,85 @@ You're free to define any helper functions.
 -}
 
 -- some help in the beginning ;)
+data Chest = Chest
+    { chestGold     :: Int
+    , chestTreasure :: Int
+    }
+    deriving (Show)
+
+add :: Chest -> Chest -> Chest
+add c1 c2 = Chest (chestGold c1 + chestGold c2)
+                  (chestTreasure c1 + chestTreasure c2)
+
 data Knight = Knight
-    { knightHealth    :: Int
-    , knightAttack    :: Int
-    , knightEndurance :: Int
+    { knightHealth     :: Int
+    , knightAttack     :: Int
+    , knightEndurance  :: Int
+    , knightExperience :: Int
+    , knightChest      :: Chest
     }
+    deriving (Show)
 
-dragonFight = error "TODO"
+data Colour
+    = Red
+    | Black
+    | Green
+    deriving (Show, Eq)
+
+data Dragon = Dragon
+    { dragonColour    :: Colour
+    , dragonChest     :: Chest
+    , dragonHealth    :: Int
+    , dragonAttack    :: Int
+    , dragonEndurance :: Int
+    }
+    deriving (Show)
```

Comment:
That's a great idea! And almost the perfect solution 🔝 

There's one problems with this particular data type definition:

1. It's possible to have treasure inside green dragon:

```haskell
myDragon = Dragon
    { dragonColor = Green
    , dragonHealth = 150
    , dragonAttack = 70
    , dragonChest = Chest 10 20
    , dragonEndurance = 150
    }
```

You can improve the code by applying the technique named "Making illegal states unrepresentable"

Reply:
I'm lost in the woods on this one.
I _think_ I might want to go for something like:
```
data Dragon
    = GreenDragon
    | RedDragon
    | BlackDragon
```
But then I'm lost on how to "attach" the common "dragon traits" (health, attack, endurance) and the specific characteristics, too (colour and chest).
If I pass them in as types to the constructor, I solved nothing. That is, with a `data Dragon = GreenDragon Chest Int Int Int` I can still have a `GreenDragon` with a `Chest` full of treasure :confused: 

Even if I used record-syntax, I wouldn't still be able to _default_ a value. Clearly I'm thinking too much from a non-FP perspective :exploding_head:

Reply:
Your ["mini patterns" article](https://kowainik.github.io/posts/haskell-mini-patterns#make-illegal-states-unrepresentable) did really help here.